packages/language-server/src/plugins/typescript/ComponentInfoProvider.ts

```diff
@@ -0,0 +1,72 @@
+import { ComponentEvents } from 'svelte2tsx';
+import ts from 'typescript';
+import { isNotNullOrUndefined } from '../../utils';
+import { findContainingNode } from './features/utils';
+
+type ComponentEventInfo = ReturnType<ComponentEvents['getAll']>;
+
+export class JsOrTsComponentInfoProvider implements ComponentInfoProvider {
+    private constructor(
+        private readonly typeChecker: ts.TypeChecker,
+        private readonly classType: ts.Type
+    ) {}
+
+    getEvents(): ComponentEventInfo {
+        const symbol = this.classType.getProperty('$$events_def');
+        if (!symbol) {
+            return [];
+        }
+
+        const declaration = symbol.valueDeclaration;
+        if (!declaration) {
+            return [];
+        }
+
+        const eventType = this.typeChecker.getTypeOfSymbolAtLocation(symbol, declaration);
+
+        return eventType
+            .getProperties()
+            .map((prop) => {
+                if (!prop.valueDeclaration) {
+                    return;
+                }
+
+                return {
+                    name: prop.name,
+                    type: this.typeChecker.typeToString(
+                        this.typeChecker.getTypeOfSymbolAtLocation(prop, prop.valueDeclaration)
+                    ),
+                    doc: ts.displayPartsToString(prop.getDocumentationComment(this.typeChecker))
+                };
+            })
+            .filter(isNotNullOrUndefined);
+    }
+
+    static create(lang: ts.LanguageService, def: ts.DefinitionInfo): ComponentInfoProvider | null {
+        const program = lang.getProgram();
+        const sourceFile = program?.getSourceFile(def.fileName);
+
+        if (!program || !sourceFile) {
+            return null;
+        }
+
+        const defClass = findContainingNode(sourceFile, def.textSpan, ts.isClassDeclaration);
+
+        if (!defClass) {
+            return null;
+        }
+
+        const typeChecker = program.getTypeChecker();
+        const classType = typeChecker.getTypeAtLocation(defClass);
+
+        if (!classType) {
+            return null;
+        }
+
+        return new JsOrTsComponentInfoProvider(typeChecker, classType);
+    }
+}
```

Comment:
What do you think about making the instance of this a property on TsOrJsDocumentSnapshot? Then it could be cached and the cache would only be deleted on updates to the snapshot.

Reply:
Not sure about the performance implications, but could this be something we could also use for the svelte2tsx-components? Then we could ditch our own implementation of creating the "events" property on the svelte2tsx return type.

Reply:
Cached in the snapshot seems like a good idea. 

> could this be something we could also use for the svelte2tsx-components

Yes. Tried with svelte2tsx-component. The biggest difference between this and the svelte2tsx approach is the fallback Event in the svelte2tsx.